framework/src/ble_manager/ble_queue.c

```diff
@@ -0,0 +1,204 @@
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <errno.h>
+#include <ble_manager/ble_queue.h>
+
+#define BLE_QUEUE_CHECK do { if (q_grp == NULL) {return BLE_QUEUE_NOT_INIT;}} while(0)
+
+static ble_queue_group *q_grp = NULL;
+static int g_run_process = 0;
+
+static int _ble_evt_handler(int argc, char *argv[])
+{
+	int priority;
+	int err_no;
+	ble_queue *q;
+
+	while (g_run_process) {
+		if (sem_wait(&q_grp->countsem) < 0) {
+			if (errno == EINTR) {
+				continue;
+			} else {
+				g_run_process = 0;
+				return -1;
+			}
+		}
+		for (priority = 0; priority < BLE_QUEUE_EVT_PRI_MAX; priority++) {
+			q = &q_grp->q[priority];
+			if (sem_trywait(&q->countsem) < 0) {
+				err_no = get_errno();
+				if (err_no == EDEADLK || err_no == EINVAL) {
+					g_run_process = 0;
+					return -1;
+				}
+			} else {
+				if (q->read_index != q->write_index) {
+					/* This Priority Queue is not Empty */
+					q_grp->caller(priority, q->queue + (q->data_size * q->read_index));
+					q->read_index = (q->read_index + 1) % q->size;
+				}
+				break;
+			}
+		}
+	}
+
+	return 0;
+}
+
+ble_queue_ret_e ble_queue_init(queue_event_caller caller)
+{
+	if (caller == NULL) {
+		return BLE_QUEUE_INVALID_ARGS;
+	}
+
+	if (q_grp == NULL) {
+		q_grp = (ble_queue_group *)zalloc(sizeof(ble_queue_group));
+		if (q_grp == NULL) {
+			return BLE_QUEUE_MEM_ALLOC_FAIL;
+		}
+		sem_init(&q_grp->countsem, 0, 0);
+		q_grp->caller = caller;
+		g_run_process = 1;
+
+		int tid = task_create("ble_evt_handler", 100, 2048, (main_t)_ble_evt_handler, NULL);
+		if (tid < 0) {
+			printf("[BLEQUEUE] create task fail\n");
+			q_grp->caller = NULL;
+			g_run_process = 0;
+			sem_post(&q_grp->countsem);
+			sem_destroy(&q_grp->countsem);
+			free(q_grp);
+			q_grp = NULL;
+			return BLE_QUEUE_FAIL;
+		}
+	}
+	return BLE_QUEUE_SUCCESS;
+}
+
+ble_queue_ret_e ble_queue_deinit(void)
+{
+	BLE_QUEUE_CHECK;
+
+	int i;
+	ble_queue_ret_e ret;
+	if (q_grp != NULL) {
+		for (i = 0; i < BLE_QUEUE_EVT_PRI_MAX; i++) {
+			ret = ble_queue_pri_clear(i);
+			if (ret != BLE_QUEUE_SUCCESS) {
+				return BLE_QUEUE_FAIL;
+			}
+		}
+		q_grp->caller = NULL;
+		g_run_process = 0;
+		sem_post(&q_grp->countsem);
+		sem_destroy(&q_grp->countsem);
+		free(q_grp);
+		q_grp = NULL;
+	}
+	return BLE_QUEUE_SUCCESS;
+}
+
+ble_queue_ret_e ble_queue_pri_set(int priority, int queue_size, int data_size) {
+	BLE_QUEUE_CHECK;
+
+	if (priority < 0 || priority >= BLE_QUEUE_EVT_PRI_MAX) {
+		return BLE_QUEUE_INVALID_ARGS;
+	}
+
+	ble_queue *q = &q_grp->q[priority];
+
+	if (q->queue == NULL) {
+		q->queue = (void *)malloc(data_size * queue_size);
+		if (q->queue == NULL) {
+			printf("[BLEQUEUE] fail to set priority queue[%d]\n", priority);
+			return BLE_QUEUE_MEM_ALLOC_FAIL;
+		}
+		sem_init(&q->countsem, 0, 0);
+		q->grp_count = &q_grp->countsem;
```

Comment:
Could you tell me why named a value countsem? I'll helpful if you explained the reason.

Reply:
The countsem of queue is for counting queue data.
Furthermore, the queue_group counts the total quantity of priority queues.